pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java

```diff
@@ -283,6 +289,109 @@ public ConfigSuccessResponse addTable(String tableConfigStr,
+        LOGGER.info("[copyTable] Successfully fetched and tweaked table config for table: {}, tableConfig: {}",
+            tableName, realtimeTableConfig.toString());
+
+        URI watermarkUri = new URI(sourceControllerUri + "/tables/" + tableName + "/consumerWatermarks");
+        SimpleHttpResponse watermarkResponse = HttpClient.wrapAndThrowHttpException(
+            HttpClient.getInstance().sendGetRequest(watermarkUri, requestHeaders));
+        String watermarkJson = watermarkResponse.getResponse();
+        LOGGER.info("[copyTable] Fetched watermarks for table: {}. Result: {}", tableName, watermarkJson);
+        WatermarkInductionResult watermarkInductionResult =
+            JsonUtils.stringToObject(watermarkJson, WatermarkInductionResult.class);
+
+        List<PartitionGroupInfo> partitionGroupInfos = watermarkInductionResult.getWatermarks().stream()
+            .map(PartitionGroupInfo::from)
+            .collect(Collectors.toList());
+
+        _pinotHelixResourceManager.addTable(realtimeTableConfig, partitionGroupInfos);
+      }
+      LOGGER.info("[copyTable] Finished Table Config copy: {}", tableName);
+      return new CopyTableResponse("success");
+    } catch (Exception e) {
+      LOGGER.error("[copyTable] Error copying table: {}", tableName, e);
+      throw new ControllerApplicationException(LOGGER, "Error copying table: " + e.getMessage(),
+          Response.Status.INTERNAL_SERVER_ERROR, e);
+    }
+  }
+
+  /**
+   * Tweaks the realtime table config with the given broker and server tenants.
+   *
+   * @param realtimeTableConfigNode The JSON object representing the realtime table config.
+   * @param brokerTenant The broker tenant to set in the config.
+   * @param serverTenant The server tenant to set in the config.
+   */
+  @VisibleForTesting
+  static void tweakRealtimeTableConfig(ObjectNode realtimeTableConfigNode, String brokerTenant, String serverTenant,
+      @Nullable Map<String, String> tagPoolReplacementMap) {
+    ObjectNode tenantConfig = (ObjectNode) realtimeTableConfigNode.get("tenants");
+    tenantConfig.put("broker", brokerTenant);
+    tenantConfig.put("server", serverTenant);
+    if (tagPoolReplacementMap == null || tagPoolReplacementMap.isEmpty()) {
+      return;
+    }
+    JsonNode instanceAssignmentConfigMap = realtimeTableConfigNode.get("instanceAssignmentConfigMap");
+    if (instanceAssignmentConfigMap == null) {
+      return;
+    }
+    java.util.Iterator<Map.Entry<String, JsonNode>> iterator = instanceAssignmentConfigMap.fields();
+    while (iterator.hasNext()) {
+      Map.Entry<String, JsonNode> entry = iterator.next();
+      JsonNode instanceAssignmentConfig = entry.getValue();
+      // tagPoolConfig is a required field
+      ObjectNode tagPoolConfig = (ObjectNode) instanceAssignmentConfig.get("tagPoolConfig");
+      // tag is a required json field
+      if (tagPoolConfig != null) {
+        String srcTag = tagPoolConfig.get("tag").asText();
```

Comment:
add a null check for srcTags.

Reply:
https://sourcegraph.com/github.com/apache/pinot@2962dd445fbd25e70d627afca14c96e4b50df5fd/-/blob/pinot-spi/src/main/java/org/apache/pinot/spi/config/table/assignment/InstanceTagPoolConfig.java?L49

tag is required in the TagPoolConfig constructor. Should be safe to skip the null check.